helpers/browser/driver.js

```diff
@@ -74,48 +67,92 @@ class ChromeProtocol {
     }, 15 * 1000);
   }
 
-  subscribeToServiceWorkerDetails(cb, resolve) {
-    var chrome = this.instance();
+  subscribeToServiceWorkerDetails(fn) {
+    var chrome = this._instance;
+
+    return new Promise(function(res, rej) {
+      chrome.ServiceWorker.enable();
+      // chrome.on("ServiceWorker.workerCreated", log)
+      // chrome.on("ServiceWorker.workerRegistrationUpdated", log)
+      chrome.on("ServiceWorker.workerVersionUpdated", data => {
+        res(fn(data));
+      });
+    });
+  }
+
+  static getEvaluationContextFor(url, instance) {
+    return new Promise((res, rej) => {
+      var errorTimeout = setTimeout((_ => rej(new Error(`No Evaluation context found for ${url}`))), 4000);
 
-    chrome.ServiceWorker.enable();
-    // chrome.on("ServiceWorker.workerCreated", log)
-    // chrome.on("ServiceWorker.workerRegistrationUpdated", log)
-    chrome.on("ServiceWorker.workerVersionUpdated", data => {
-      cb(data, resolve);
+      instance.on('Runtime.executionContextCreated', evalContext => {
+        if (evalContext.context.origin.indexOf(url) !== -1) {
+          clearTimeout(errorTimeout);
+          res(evalContext.context.id);
+        }
+      });
     });
   }
 
-  evaluateScript(chrome) { /* TODO(paulirish): scriptStr parameter */
-    return new Promise(function(resolve, reject) {
-      chrome.Runtime.evaluate({
-        expression: 'alert(navigator.userAgent)',
-        contextId: 0
-      }, resolve);
+  evaluateScript(scriptStr, url) {
+    var chrome = this._instance;
+    var wrappedScriptStr = '(' + scriptStr.toString() + ')()';
+
+    chrome.Runtime.enable();
+
+    return ChromeProtocol.getEvaluationContextFor(url, chrome)
+    .then(contextId => {
+      return new Promise((resolve, reject) => {
+        function evalInContext(){
+
+          chrome.Runtime.evaluate({
+            expression: wrappedScriptStr,
+            contextId: contextId,
+          }, (err, evalRes) => {
+            if (err || evalRes.wasThrown){
+              return reject(evalRes);
+            }
+
+            chrome.Runtime.getProperties({
+              objectId : evalRes.result.objectId
+            }, (err, res) => {
+              evalRes.result.props = {};
+              if (Array.isArray(res.result)) {
+                res.result.forEach(prop => {
+                  evalRes.result.props[prop.name] = prop.value ? prop.value.value : prop.get.description;
+                });
+              }
+              resolve(evalRes.result);
+            })
+          });
+        }
+        // allow time to pull in some contexts
+        setTimeout(evalInContext, 500);
+      });
     });
   }
 
-  profilePageLoad(chrome) {
+  profilePageLoad(url) {
     return new Promise((function(resolve, reject) {
       var rawEvents = [];
       this.resetFailureTimeout(reject);
 
-      chrome.Page.enable();
-      chrome.Tracing.start({
+      this._instance.Page.enable();
+      this._instance.Tracing.start({
```

Comment:
may be more a question for @paulirish, but is there an `error` event we should be listening for here to `reject` on?